File: lib/dal/freelist.ml
(** This component is part of the DAL. *)

open Utils

(** Manages the manages free and used pages.
    [max_page]: holds the latest page num allocated.
    [released_pages]: holds all the ids that were released during delete
    [max_page] is incremented and a new page is created thus increasing the file size. *)
type t =
  { mutable max_page : Page_num.t;
    mutable released_pages : Page_num.t array }

let make () = {max_page = Page.meta; released_pages = [||]}

(** If possible, fetch pages first from the released pages. Else, increase the maximum page *)
let get_next_page (freelist : t) =
  if Array.length freelist.released_pages <> 0 then (
    let page_id = freelist.released_pages |> Array.last in
      freelist.released_pages
      <- freelist.released_pages
         |> Array.sub ~pos:0 ~len:(Array.length freelist.released_pages - 1);
      page_id
  ) else (
    freelist.max_page <- Page_num.(freelist.max_page + 1I);
    freelist.max_page
  )
;;

let release_page (freelist : t) (page : Page_num.t) =
  freelist.released_pages <- Array.append freelist.released_pages [|page|]
;;

File: Toocal/Freelist.fs
module Toocal.Core.DataAccessLayer.Freelist

open System
open Toocal.Core.DataAccessLayer.Page
open Toocal.Core.DataAccessLayer
open Toocal.Core.ISerializable

/// Which pages are free and which are occupied.
/// Pages can also be freed if they become empty,
/// so we need to reclaim them for future use to avoid fragmentation.
type Freelist() =

  /// Holds the maximum page allocated. _MaxPage*PageSize = fileSize
  let mutable _MaxPage: PageNum = 0UL

  /// Pages that were previouslly allocated but are now free
  let _ReleasedPages = new Collections.Generic.Stack<PageNum>()

  member this.MaxPage
    with get () = _MaxPage
    and set (maxPage: PageNum) = _MaxPage <- maxPage

  member this.ReleasedPages = _ReleasedPages

  /// If possible, fetch pages first from the released pages.
  /// Else, increase the maximum page
  member public this.NextPage() =
    if _ReleasedPages.Count <> 0 then
      _ReleasedPages.Pop()
    else
      _MaxPage <- _MaxPage + 1UL
      _MaxPage

  member public this.ReleasePage(page: PageNum) = _ReleasedPages.Push page

  member this.Serialize = (this :> ISerializable).Serialize
  member this.Deserialize = (this :> ISerializable).Deserialize

  interface ISerializable with
    member this.Serialize(buffer: Byte[]) =
      let mutable pos = 0
      let serializedMaxPage = this.MaxPage |> uint16 |> BitConverter.GetBytes

      let serializedPageCount =
        this.ReleasedPages.Count |> uint16 |> BitConverter.GetBytes

      Array.blit serializedMaxPage 0 buffer pos serializedMaxPage.Length
      pos <- pos + 2

      Array.blit serializedPageCount 0 buffer pos serializedPageCount.Length
      pos <- pos + 2

      let mutable page = this.ReleasedPages.GetEnumerator()

      while page.MoveNext() do
        let serializedPage = BitConverter.GetBytes page.Current
        Array.blit serializedPage 0 buffer pos serializedPage.Length
        pos <- pos + Page.SIZE

    member this.Deserialize(buffer: Byte[]) =
      let mutable pos = 0
      this.MaxPage <- BitConverter.ToUInt16(buffer[pos..]) |> uint64
      pos <- pos + 2

      let releasedPagesCount = BitConverter.ToUInt16(buffer[pos..]) |> uint64
      pos <- pos + 2

      for _ = 0 to (BitConverter.ToUInt16 buffer[pos..] |> int) - 1 do
        buffer[pos..] |> BitConverter.ToUInt64 |> this.ReleasedPages.Push
        pos <- pos + Page.SIZE
